funcx_sdk/funcx/tests/unit/test_executor.py

```diff
@@ -0,0 +1,12 @@
+from funcx import FuncXExecutor
+
+
+def test_executor_shutdown(mocker):
+    mocker.patch("funcx.sdk.executor.atexit")
+    mocker.patch("funcx.sdk.executor.ExecutorPollerThread.event_loop_thread")
+    fcli = mocker.MagicMock()
+    fexe = FuncXExecutor(funcx_client=fcli)
+    fexe.poller_thread.atomic_controller.increment()
+    fexe.poller_thread.atomic_controller.decrement()
```

Comment:
What happens as a result of the `increment(); decrement()` operations?

I see that the executor contains the poller thread, and the poller thread sets up with counter/callback mechanism:
```
self.atomic_controller = AtomicController(self.start, noop)
```
above, we have `def noop(): return`, so it seems like incrementing starts, but decrementing does nothing. Is that right?

Then it would seem that below in this test, `fexe.shutdown()` is checking the case when `fexe._ws is not None`.

Is that the case we want to test? My read on this fix is that it's fixing the case when `fexe._ws is None`. So we should ideally test both cases. I'm still reading and trying to figure out how to trigger "clean shutdown" in the tests.

Reply:
The core of this is indeed to not fail when `_ws is None`.  For the specifics of this test, `decrement()` is not necessary (`noop()`, as you noted) -- that's more just to show that the test is in good faith.  The shutdown occurs in `ws_polling_task.py`:

```python
async def set_result(self, task_id, data, pending_futures):
  # ...
  if self.atomic_controller is not None:
    count = self.atomic_controller.decrement()
    # Only close when count == 0 and unknown_results are empty
    if count == 0 and len(self.unknown_results) == 0:
      await self.ws.close()
      self._ws = None
      return True
```

Ick!

Reply:
I agree, ick! indeed.
But I also see that there's `async def close` as a method of the WebSocketPollingTask.

So if we're using `pytest-asyncio`, it should be possible to write an async test which does `await fexe.ws_handler.close()`.

Or will that fail for some reason?